# Metadata
- This notebook is for processing the various raw information scraped and combining it into one `metadata.csv`
- For each data source, the file will be stored at `data/raw/{data source}/metadata.csv`

In [1]:
import pandas as pd
from tqdm import tqdm

import os
import requests

tqdm.pandas()

In [2]:
# change working directory to the project root directory
current_dir = os.getcwd()
os.chdir(current_dir + "/../../")
# this should be the project root directory
os.getcwd()

'/Users/gustavosantos/code/omdena/SaoPauloBrazilChapter_BrazilianSignLanguage'

## INES

- Uses the existing columns in `INES_Metadata.csv`
- checks if the video file exists or not
- adds the column `file_exists`

Load the `INES_metadata.csv` to get the Video URLs

In [ ]:
ines_csv = pd.read_csv("data/raw/INES/INES_Metadata.csv")
ines_csv.head()

Letter     Word                                          Video URL  \
0      A        A  https://www.ines.gov.br/dicionario-de-libras/p...   
1      A  ABACATE  https://www.ines.gov.br/dicionario-de-libras/p...   
2      A  ABACAXI  https://www.ines.gov.br/dicionario-de-libras/p...   
3      A   ABAFAR  https://www.ines.gov.br/dicionario-de-libras/p...   
4      A   ABAIXO  https://www.ines.gov.br/dicionario-de-libras/p...   

                                           Image URL Assuntos  \
0  https://www.ines.gov.br/https://www.ines.gov.b...   NENHUM   
1  https://www.ines.gov.br/https://www.ines.gov.b...    FRUTA   
2  https://www.ines.gov.br/https://www.ines.gov.b...    FRUTA   
3  https://www.ines.gov.br/https://www.ines.gov.b...   NENHUM   
4  https://www.ines.gov.br/https://www.ines.gov.b...   NENHUM   

                                             Acepção  \
0  Primeira letra do alfabeto da língua portugues...   
1  O fruto do abacateiro. Comestível, tem a polpa...   
2  Fruta de casca grossa e áspera. Sua polpa pode...   
3             Cobrir ou fechar, para manter o calor.   
4  Lugar, posição ou situação inferior, em relaçã...   

                                             Exemplo  \
0  Invente qualquer palavra que comece com a letr...   
1                   Você gosta de abacate com leite?   
2      Hoje tomei suco de abacaxi, ele estava ácido.   
3  Se você quer abafar seu quarto, é melhor fecha...   
4  Não é no primeiro apartamento abaixo, é no seg...   

                                      Exemplo Libras Classe Gramatical  \
0          VOCÊ INVENTAR QUALQUER PALAVRA COMEÇAR A.       SUBSTANTIVO   
1                   VOCÊ GOSTAR ABACATE LEITE JUNTO?       SUBSTANTIVO   
2                  HOJE S-U-C-O ABACAXI BEBER ÁCID@.       SUBSTANTIVO   
3  S-I VOCÊ QUERER QUARTO SE@ ABAFAR A-R? MELHOR ...             VERBO   
4           APARTAMENTO PRIMEIR@ NÃO SEGUND@ ABAIXO.              ADV.   

     Origem  
0  Nacional  
1  Nacional  
2  Nacional  
3  Nacional  
4  Nacional

Find a difference in response between URLs for words that have/don't have a video file 

In [ ]:
# invalid url
invalid = "https://www.ines.gov.br/dicionario-de-libras/public/media/palavras/videos/baba1Sm_Prog001.mp4"
response = requests.get(invalid, stream=True)
response.headers["Content-Type"]

'text/html;charset=utf-8'

In [ ]:
# valid url
valid = "https://www.ines.gov.br/dicionario-de-libras/public/media/palavras/videos/babadorSm_Prog001.mp4"
response = requests.get(valid, stream=True)
response.headers["Content-Type"]

'video/mp4'

This takes a while to run since it checks every URL, so just get the info from `video_file_exists.csv` or `metadata.csv`

*(took X minutes on Ben's PC/connection)*

In [45]:
import asyncio
import aiohttp
import random
from aiohttp import ClientConnectorError, ClientTimeout

In [ ]:
video_urls = ines_csv["Video URL"].tolist()

file_exists = []

In [ ]:
# Limit concurrent requests
MAX_CONCURRENT = 20  # Reduce this if still getting errors
RETRY_COUNT = 5  # Increased from 3


async def check_url(url, index, total, semaphore, retry_count=RETRY_COUNT):
    timeout = ClientTimeout(total=30)

    async with semaphore:  # Control concurrent requests
        for attempt in range(retry_count):
            try:
                # Longer delay between requests
                await asyncio.sleep(random.uniform(0.5, 1.0))

                async with aiohttp.ClientSession(timeout=timeout) as session:
                    async with session.get(url) as response:
                        exists = response.headers["Content-Type"] == "video/mp4"
                        percent = (index + 1) / total * 100
                        if percent % 5 < (1 / total * 100):
                            print(
                                f"Progress: {percent:.1f}% ({index + 1}/{total} URLs checked)"
                            )
                        return exists

            except (ClientConnectorError, asyncio.TimeoutError) as e:
                if attempt == retry_count - 1:  # Last attempt
                    print(
                        f"\nFailed to check URL after {retry_count} attempts: {url}\nError: {str(e)}"
                    )
                    return False
                print(
                    f"\nRetrying URL {index + 1}/{total} (attempt {attempt + 2}/{retry_count})"
                )
                # Longer wait between retries
                await asyncio.sleep(random.uniform(2, 3))
            except Exception as e:
                print(f"\nUnexpected error for URL {index + 1}/{total}: {str(e)}")
                return False


async def check_all_urls(urls):
    total = len(urls)
    print(f"Starting to check {total} URLs...")

    # Create semaphore to limit concurrent requests
    semaphore = asyncio.Semaphore(MAX_CONCURRENT)

    tasks = [check_url(url, i, total, semaphore) for i, url in enumerate(urls)]
    results = await asyncio.gather(*tasks, return_exceptions=True)

    # Filter out any exceptions and count successes
    valid_results = [r for r in results if isinstance(r, bool)]
    exists_count = sum(1 for r in valid_results if r)

    print(
        f"\nFinished checking URLs. Found {exists_count} existing videos out of {total}"
    )
    return results


# Use it with:
file_exists = await check_all_urls(video_urls)

Starting to check 5819 URLs...
Progress: 5.0% (291/5819 URLs checked)
Progress: 10.0% (582/5819 URLs checked)
Progress: 15.0% (873/5819 URLs checked)
Progress: 20.0% (1164/5819 URLs checked)
Progress: 25.0% (1455/5819 URLs checked)
Progress: 30.0% (1746/5819 URLs checked)
Progress: 35.0% (2037/5819 URLs checked)
Progress: 40.0% (2328/5819 URLs checked)
Progress: 45.0% (2619/5819 URLs checked)
Progress: 50.0% (2910/5819 URLs checked)
Progress: 55.0% (3201/5819 URLs checked)
Progress: 60.0% (3492/5819 URLs checked)
Progress: 65.0% (3783/5819 URLs checked)
Progress: 70.0% (4074/5819 URLs checked)
Progress: 75.0% (4365/5819 URLs checked)
Progress: 80.0% (4656/5819 URLs checked)
Progress: 85.0% (4947/5819 URLs checked)

Retrying URL 4734/5819 (attempt 2/5)

Retrying URL 4827/5819 (attempt 2/5)
Progress: 90.0% (5238/5819 URLs checked)
Progress: 95.0% (5529/5819 URLs checked)
Progress: 100.0% (5819/5819 URLs checked)

Finished checking URLs. Found 5468 existing videos out of 5819


In [88]:
len(file_exists)

5819

Save this as `video_file_exists.csv` so that if `metadata.csv` is overwritten, we don't lose this info

In [ ]:
file_exists_df = ines_csv[["Word", "Video URL"]].copy()
file_exists_df.rename(columns={"Word": "label", "Video URL": "video_url"}, inplace=True)
file_exists_df["file_exists"] = file_exists

file_exists_df.to_csv("data/raw/INES/video_file_exists.csv", index=False)

Make `metadata_df` using the already saved `video_file_exists.csv` to avoid running the requests code again

In [ ]:
ines_csv = pd.read_csv("data/raw/INES/INES_Metadata.csv")
ines_csv.columns = ines_csv.columns.str.lower()
ines_csv.rename(columns={"word": "label", "video url": "video_url"}, inplace=True)
ines_csv.drop(columns=["image url"], inplace=True)

In [ ]:
file_exists_df = pd.read_csv("data/raw/INES/video_file_exists.csv")

In [5]:
file_exists_df.shape, ines_csv.shape

((5819, 3), (5819, 9))

In [ ]:
(
    (file_exists_df.label == ines_csv.label).sum(),
    (file_exists_df.video_url == ines_csv.video_url).sum(),
)

(5819, 5819)

In [ ]:
metadata_df = pd.concat(
    [ines_csv, file_exists_df.drop(["label", "video_url"], axis=1)], axis=1
)

In [9]:
metadata_df.columns

Index(['letter', 'label', 'video_url', 'assuntos', 'acepção', 'exemplo',
       'exemplo libras', 'classe gramatical', 'origem', 'file_exists'],
      dtype='object')

In [10]:
def number_in_label(row):
    """Returns True if there is a number in the label else False"""
    return any(char.isdigit() for char in row["label"])


def remove_number_from_label(row):
    """Removes numbers from the label"""
    return "".join([char for char in row["label"] if not char.isdigit()])

In [11]:
metadata_df["number_in_label"] = metadata_df.progress_apply(number_in_label, axis=1)
metadata_df["scraped_label"] = metadata_df["label"]
metadata_df["label"] = metadata_df.progress_apply(remove_number_from_label, axis=1)

100%|██████████| 5819/5819 [00:00<00:00, 141333.82it/s]


In [ ]:
metadata_df["signer_number"] = 0
metadata_df["data_source"] = "ne"

In [14]:
# Save the metadata.csv file
metadata_df[
    [
        "label",
        "video_url",
        "signer_number",
        "data_source",
        "scraped_label",
        "number_in_label",
        "file_exists",
        "letter",
        "assuntos",
        "acepção",
        "exemplo",
        "exemplo libras",
        "classe gramatical",
        "origem",
    ]
].to_csv("data/raw/INES/metadata.csv", index=False)

In [15]:
metadata_df.file_exists.value_counts()

file_exists
True     5468
False     351
Name: count, dtype: int64

## V-Librasil

- Uses the existing information in `v_librasil_words_n_links.txt`
- Uses the links to each word's page
- Collects the video file links for each interpreter
- Combines it into a `.csv` file

#### Fixing `v_librasil_words_n_links.txt`

In [ ]:
# load .txt file as a string
words_and_links = open(
    "data/raw/V-Librasil/words/v_librasil_words_n_links.txt", "r"
).read()
words_and_links = words_and_links.split("\n")

In [ ]:
# one line had two lines combined
for line in words_and_links:
    if len(line.split("https://")) > 2:
        print(line)

In [ ]:
# i edited the file directly to fix it
# load .txt file as a string
words_and_links = open(
    "data/raw/V-Librasil/words/v_librasil_words_n_links.txt", "r"
).read()
words_and_links = words_and_links.split("\n")

# one line had two lines combined
for line in words_and_links:
    if "Bigode" in line:
        print(line)
    if "Bilhão" in line:
        print(line)

Bigode https://libras.cin.ufpe.br/sign/826
Bilhão https://libras.cin.ufpe.br/sign/105


#### Making `metadata.csv`

turning `v_librasil_words_n_links.txt` into a DataFrame

In [ ]:
words_and_links = pd.read_csv(
    "data/raw/V-Librasil/words/v_librasil_words_n_links.txt",
    sep="https",
    header=None,
    engine="python",
)
words_and_links.columns = ["label", "sign_url"]
words_and_links.sign_url = words_and_links.sign_url.apply(lambda x: "https" + x)
words_and_links.head()

label                              sign_url
0  À noite toda    https://libras.cin.ufpe.br/sign/885
1  À tarde toda    https://libras.cin.ufpe.br/sign/100
2       Abacaxi    https://libras.cin.ufpe.br/sign/817
3        Abanar   https://libras.cin.ufpe.br/sign/1536
4     Abandonar     https://libras.cin.ufpe.br/sign/71

getting video URLs (~3 for each sign)

In [ ]:
from bs4 import BeautifulSoup


def get_video_urls(url):
    response = requests.get(url)

    if response.status_code == 200:
        video_links = []
        signer_numbers = []
        signer_order = ""
        html = response.content
        soup = BeautifulSoup(html, "html.parser")

        # go inside div class container
        container = soup.find("section", class_="page-section").find(
            "div", class_="container"
        )
        # go inside div class row
        rows = [child for child in container.children if child.name == "div"]

        # get the video and signer number
        for row in rows:
            signer_number = row.find("h2").text.strip().split(" ")[1]
            link = row.find("source").get("src")
            video_links.append(link)
            signer_numbers.append(signer_number)
            signer_order += signer_number
        return video_links, signer_numbers, signer_order

    else:
        print(f"Response code != 200, Failed to get video URLs for {link}")
        return None

In [ ]:
words_and_links["all_video_info"] = words_and_links["sign_url"].progress_apply(
    get_video_urls
)

  4%|▍         | 53/1364 [00:30<12:36,  1.73it/s]


KeyboardInterrupt: 

One by one takes a while, so I asked AI to make it async

In [8]:
df = words_and_links.copy()

In [ ]:
import pandas as pd
from tqdm import tqdm
import logging
from typing import Dict, Any


class AsyncRequestProcessor:
    def __init__(
        self,
        max_concurrent: int = 10,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        timeout: int = 30,
    ):
        self.max_concurrent = max_concurrent
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.timeout = timeout
        self.setup_logging()

    def setup_logging(self):
        logging.basicConfig(
            level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
        )
        self.logger = logging.getLogger(__name__)

    async def process_url(
        self, url: str, session: aiohttp.ClientSession, semaphore: asyncio.Semaphore
    ) -> Dict[str, Any]:
        for attempt in range(self.max_retries):
            try:
                async with semaphore:
                    # Random delay between requests
                    await asyncio.sleep(random.uniform(0.1, 0.3))

                    # Call get_video_urls function
                    video_links, signer_numbers, signer_order = await get_video_urls(
                        url, session
                    )
                    return {
                        "url": url,
                        "status": "success",
                        "video_links": video_links,
                        "signer_numbers": signer_numbers,
                        "signer_order": signer_order,
                        "attempts": attempt + 1,
                    }

            except aiohttp.ClientError as e:
                if attempt == self.max_retries - 1:
                    return {
                        "url": url,
                        "status": "error",
                        "error": f"Network error: {str(e)}",
                        "attempts": attempt + 1,
                    }
                await asyncio.sleep(self.retry_delay * (attempt + 1))
            except Exception as e:
                return {
                    "url": url,
                    "status": "error",
                    "error": str(e),
                    "attempts": attempt + 1,
                }

    async def process_urls(self, urls: list) -> list:
        semaphore = asyncio.Semaphore(self.max_concurrent)
        timeout = aiohttp.ClientTimeout(total=self.timeout)

        async with aiohttp.ClientSession(timeout=timeout) as session:
            tasks = [self.process_url(url, session, semaphore) for url in urls]

            results = []
            failed = 0
            for f in tqdm(
                asyncio.as_completed(tasks), total=len(tasks), desc="Processing URLs"
            ):
                result = await f
                if result["status"] == "error":
                    failed += 1
                    self.logger.warning(
                        f"Failed to process {result['url']}: {result['error']}"
                    )
                results.append(result)

            self.logger.info(
                f"Completed processing {len(results)} URLs. Failed: {failed}"
            )
            return results


async def get_video_urls(url: str, session: aiohttp.ClientSession):
    async with session.get(url) as response:
        if response.status == 200:
            video_links = []
            signer_numbers = []
            signer_order = ""

            # Get the HTML content
            html = await response.text()
            soup = BeautifulSoup(html, "html.parser")

            # go inside div class container
            container = soup.find("section", class_="page-section").find(
                "div", class_="container"
            )
            # go inside div class row
            rows = [child for child in container.children if child.name == "div"]

            # get the video and signer number
            for row in rows:
                signer_number = row.find("h2").text.strip().split(" ")[1]
                link = row.find("source").get("src")
                video_links.append(link)
                signer_numbers.append(signer_number)
                signer_order += signer_number
            return video_links, signer_numbers, signer_order
        else:
            raise aiohttp.ClientError(
                f"Response code {response.status}, Failed to get video URLs for {url}"
            )


# For Jupyter notebook execution
async def main(df):
    processor = AsyncRequestProcessor(
        max_concurrent=10, max_retries=3, retry_delay=1.0, timeout=30
    )

    results = await processor.process_urls(df["sign_url"].tolist())
    return pd.DataFrame(results)


# Execute in Jupyter
try:
    # Get current notebook's event loop
    loop = asyncio.get_event_loop()
    results_df = await main(df)  # Use await here in Jupyter

    # Process results into final dataframe
    successful_results = results_df[results_df["status"] == "success"]

    # Create separate columns for video links and signer information
    df_expanded = pd.concat(
        [
            df,
            pd.DataFrame(
                {
                    "video_links": successful_results["video_links"],
                    "signer_numbers": successful_results["signer_numbers"],
                    "signer_order": successful_results["signer_order"],
                }
            ),
        ],
        axis=1,
    )

except Exception as e:
    logging.error(f"Processing failed: {str(e)}")

Processing URLs: 100%|██████████| 1364/1364 [03:20<00:00,  6.79it/s]
2025-03-06 17:12:54,283 - INFO - Completed processing 1364 URLs. Failed: 0


In [10]:
results_df.head()

url   status  \
0  https://libras.cin.ufpe.br/sign/700  success   
1  https://libras.cin.ufpe.br/sign/990  success   
2  https://libras.cin.ufpe.br/sign/438  success   
3   https://libras.cin.ufpe.br/sign/77  success   
4  https://libras.cin.ufpe.br/sign/551  success   

                                         video_links signer_numbers  \
0  [https://libras.cin.ufpe.br/storage/videos/202...      [3, 1, 2]   
1  [https://libras.cin.ufpe.br/storage/videos/202...      [3, 1, 2]   
2  [https://libras.cin.ufpe.br/storage/videos/202...      [1, 2, 3]   
3  [https://libras.cin.ufpe.br/storage/videos/202...      [1, 2, 3]   
4  [https://libras.cin.ufpe.br/storage/videos/202...      [1, 2, 3]   

  signer_order  attempts  
0          312         1  
1          312         1  
2          123         1  
3          123         1  
4          123         1

In [11]:
results_df.status.value_counts()

status
success    1364
Name: count, dtype: int64

In [ ]:
results_df = results_df.explode(["video_links", "signer_numbers"])
results_df.head()

url   status  \
0  https://libras.cin.ufpe.br/sign/700  success   
0  https://libras.cin.ufpe.br/sign/700  success   
0  https://libras.cin.ufpe.br/sign/700  success   
1  https://libras.cin.ufpe.br/sign/990  success   
1  https://libras.cin.ufpe.br/sign/990  success   

                                         video_links signer_numbers  \
0  https://libras.cin.ufpe.br/storage/videos/2021...              3   
0  https://libras.cin.ufpe.br/storage/videos/2021...              1   
0  https://libras.cin.ufpe.br/storage/videos/2021...              2   
1  https://libras.cin.ufpe.br/storage/videos/2021...              3   
1  https://libras.cin.ufpe.br/storage/videos/2021...              1   

  signer_order  attempts  
0          312         1  
0          312         1  
0          312         1  
1          312         1  
1          312         1

In [13]:
words_and_links.head()

label                              sign_url
0  À noite toda    https://libras.cin.ufpe.br/sign/885
1  À tarde toda    https://libras.cin.ufpe.br/sign/100
2       Abacaxi    https://libras.cin.ufpe.br/sign/817
3        Abanar   https://libras.cin.ufpe.br/sign/1536
4     Abandonar     https://libras.cin.ufpe.br/sign/71

In [ ]:
# join back to words_and_links on sign_url
results_df.rename(
    columns={
        "url": "sign_url",
        "video_links": "video_url",
        "signer_numbers": "signer_number",
    },
    inplace=True,
)
metadata_df = words_and_links.merge(results_df, on="sign_url", how="left")
metadata_df.head()

label                             sign_url   status  \
0  À noite toda   https://libras.cin.ufpe.br/sign/885  success   
1  À noite toda   https://libras.cin.ufpe.br/sign/885  success   
2  À noite toda   https://libras.cin.ufpe.br/sign/885  success   
3  À tarde toda   https://libras.cin.ufpe.br/sign/100  success   
4  À tarde toda   https://libras.cin.ufpe.br/sign/100  success   

                                           video_url signer_number  \
0  https://libras.cin.ufpe.br/storage/videos/2021...             1   
1  https://libras.cin.ufpe.br/storage/videos/2021...             3   
2  https://libras.cin.ufpe.br/storage/videos/2021...             2   
3  https://libras.cin.ufpe.br/storage/videos/2020...             1   
4  https://libras.cin.ufpe.br/storage/videos/2020...             2   

  signer_order  attempts  
0          132         1  
1          132         1  
2          132         1  
3          123         1  
4          123         1

In [15]:
metadata_df.columns

Index(['label', 'sign_url', 'status', 'video_url', 'signer_number',
       'signer_order', 'attempts'],
      dtype='object')

In [ ]:
metadata_df["data_source"] = "vl"

Save selected columns

In [ ]:
metadata_df[
    ["label", "video_url", "signer_number", "data_source", "sign_url", "signer_order"]
].to_csv("data/raw/V-Librasil/metadata.csv", index=False)

## SignBank

### Explore the `full_metadata_SignBank.csv` file

In [ ]:
sb_csv = pd.read_csv("data/raw/SignBank/full_metadata_SignBank.csv")
sb_csv.rename(
    columns={
        "Alphabet Label": "letter",
        "Sign Label": "label",
        "Video URL": "video_url",
    },
    inplace=True,
)
sb_csv.head()

letter      label                                          video_url
0      A      À-TOA  https://levantelab.storage.googleapis.com/libr...
1      A    À-VISTA  https://levantelab.storage.googleapis.com/libr...
2      A    ABACAXI  https://levantelab.storage.googleapis.com/libr...
3      A     ABANAR  https://levantelab.storage.googleapis.com/libr...
4      A  ABANDONAR  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...

In [ ]:
sb_csv["video_url_root"] = (
    sb_csv["video_url"].str.split("https://").str[1].str.split(".").str[0]
)
sb_csv["video_url_root"].value_counts()

video_url_root
videos        2992
objetos         63
levantelab      28
Name: count, dtype: int64

In [ ]:
sb_csv["video_url_ext"] = sb_csv["video_url"].str.split(".").str[-1]
sb_csv["video_url_ext"].value_counts()

video_url_ext
mp4              3062
mov                21
URL_NOT_FOUND       1
Name: count, dtype: int64

In [ ]:
sb_csv[sb_csv["video_url_ext"] == "mov"].video_url_root.value_counts()

video_url_root
objetos    21
Name: count, dtype: int64

In [ ]:
sb_csv[sb_csv["video_url_root"] == "objetos"].video_url_ext.value_counts()

video_url_ext
mp4    42
mov    21
Name: count, dtype: int64

In [ ]:
for i in [4, 3, 445]:
    print(sb_csv.iloc[i]["label"])
    print(sb_csv.iloc[i]["video_url_root"])
    print(sb_csv.iloc[i]["video_url"])
    print()

ABANDONAR
videos
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/ABANDONAR.mp4

ABANAR
levantelab
https://levantelab.storage.googleapis.com/libras/gerais/swiQhJZ6FoTLXBi1u8AgLBLjvhzUO1fbPLK1EEao.mp4

BOMBOM-2
objetos
https://objetos.sites.ufsc.br/libras/gerais/BaJ58VuGYbwzgx40LL8JhLfHLCAHVT2Uv4DLeXVt.mp4



### video_url validity (based on manual checking of a sample)

#### *videos.nals* - 2992 entries

URLs starting with `videos.nals...` are valid.

They correspond to videos from a signer with:
- facial hair
- an earring
- a black button-up shirt

They follow the URL format:
>`https://videos.nals.cce.ufsc.br/SignBank/Vídeos/{label}.mp4`

The web URL can handle accented letters. The í in Videos is accented. The {label} is sometimes accented.

#### *objetos* - 63 entries

URLs starting with `objetos...` are valid.

They correspond to videos from a signer with:
- no facial hair
- a black t-shirt

They follow the URL format:
`https://objetos.sites.ufsc.br/libras/gerais/{40-character-code}.{file-extension}`
- where `40-character-code` is a 40 character code (possibly derived from the label?) like:
    - `HWQBq7whvIWSxzN3HHii6xk0Npu0qUTxJBZiMPeI`
    - `I021pwWFtHqpgspqKCpiS5Q8quymB9urO9wfJCru`
- and `file-extension` is either `.mp4` or `.mov`
    - 42 end in `.mp4`, which opens the video in browser
    - 21 end in `.mov`, which starts a direct download


#### *levantelab* - 28 entries
URLs starting with `levantelab.storage...` are invalid and return:
>This XML file does not appear to have any style information associated with it. The document tree is shown below.
>```
><Error>
><Code>UserProjectAccountProblem</Code>
><Message>The project to be billed is associated with a closed billing account.</Message>
><Details>The billing account for the owning project is disabled in state closed</Details>
></Error>
>```

The videos don't load on the SignBank portal either, so the issue is on their end. 

They follow the URL format:
`https://levantelab.storage.googleapis.com/libras/gerais/{40-character-code}.mp4`

Copying the `{40-character-code}` into the *objetos* URL format **DOES NOT** work:
>...\
>```<Message>The specified key does not exist.</Message>```\
>...

Copying the label into the *videos.nals* URL format **DOES** work:
- ABANAR: https://levantelab.storage.googleapis.com/libras/gerais/swiQhJZ6FoTLXBi1u8AgLBLjvhzUO1fbPLK1EEao.mp4
    - `INVALID`
- ABANAR: https://videos.nals.cce.ufsc.br/SignBank/Vídeos/{ABANAR}.mp4
    - `VALID`

### Fixing URLs

In [ ]:
sb_csv.rename(columns={"video_url": "scraped_url"}, inplace=True)
sb_csv["fixed_url"] = sb_csv["scraped_url"]

Something happened with `FOLGAR`, as the video on the SignBank portal is availabe, and the URL `https://videos.nals.cce.ufsc.br/SignBank/Vídeos/FOLGAR.mp4` works

In [ ]:
sb_csv[sb_csv["video_url_ext"] == "URL_NOT_FOUND"]

letter   label    scraped_url video_url_root  video_url_ext  \
1248      F  FOLGAR  URL_NOT_FOUND            NaN  URL_NOT_FOUND   

          fixed_url  
1248  URL_NOT_FOUND

In [ ]:
# manually change entries at label == 'FOLGAR'
index = sb_csv[sb_csv["label"] == "FOLGAR"].index
sb_csv.loc[index, "fixed_url"] = (
    "https://videos.nals.cce.ufsc.br/SignBank/Vídeos/FOLGAR.mp4"
)
sb_csv.loc[index, "video_url_root"] = "videos"
sb_csv.loc[index, "video_url_ext"] = "mp4"

In [ ]:
sb_csv[sb_csv["label"] == "FOLGAR"]

letter   label    scraped_url video_url_root video_url_ext  \
1248      F  FOLGAR  URL_NOT_FOUND         videos           mp4   

                                              fixed_url  
1248  https://videos.nals.cce.ufsc.br/SignBank/Vídeo...

Fix the 28 `levantelab` entries

In [ ]:
sb_csv[sb_csv["video_url_root"] == "levantelab"]

letter                        label  \
0         A                        À-TOA   
1         A                      À-VISTA   
2         A                      ABACAXI   
3         A                       ABANAR   
5         A                       ABELHA   
6         A                     ABENÇOAR   
7         A                    ABORDAGEM   
10        A                      ABRAÇAR   
96        A                     ALEMANHA   
284       A                        ABRIR   
285       A                       ACUSAR   
286       A                     AGUENTAR   
287       A                      ABACATE   
288       A                    ABACAXI-2   
289       A                    ABACAXI-3   
290       A                      ABAIXAR   
1152      E  ESTADOS-UNIDOS-DA-AMÉRICA-2   
1540      I                      ICÔNICO   
1903      M                  MISSIONÁRIO   
1904      M                     MAIONESE   
2049      O                    OLÍMPICOS   
2355      P                   PARÂMETROS   
2356      P                      PRA-QUÊ   
2385      Q                 QUINTA-FEIRA   
2722      S                  SOBRE-O-QUÊ   
3053      V                  VEGETARIANO   
3054      V                     VICIAR-2   
3055      V                  VELOCÍMETRO   

                                            scraped_url video_url_root  \
0     https://levantelab.storage.googleapis.com/libr...     levantelab   
1     https://levantelab.storage.googleapis.com/libr...     levantelab   
2     https://levantelab.storage.googleapis.com/libr...     levantelab   
3     https://levantelab.storage.googleapis.com/libr...     levantelab   
5     https://levantelab.storage.googleapis.com/libr...     levantelab   
6     https://levantelab.storage.googleapis.com/libr...     levantelab   
7     https://levantelab.storage.googleapis.com/libr...     levantelab   
10    https://levantelab.storage.googleapis.com/libr...     levantelab   
96    https://levantelab.storage.googleapis.com/libr...     levantelab   
284   https://levantelab.storage.googleapis.com/libr...     levantelab   
285   https://levantelab.storage.googleapis.com/libr...     levantelab   
286   https://levantelab.storage.googleapis.com/libr...     levantelab   
287   https://levantelab.storage.googleapis.com/libr...     levantelab   
288   https://levantelab.storage.googleapis.com/libr...     levantelab   
289   https://levantelab.storage.googleapis.com/libr...     levantelab   
290   https://levantelab.storage.googleapis.com/libr...     levantelab   
1152  https://levantelab.storage.googleapis.com/libr...     levantelab   
1540  https://levantelab.storage.googleapis.com/libr...     levantelab   
1903  https://levantelab.storage.googleapis.com/libr...     levantelab   
1904  https://levantelab.storage.googleapis.com/libr...     levantelab   
2049  https://levantelab.storage.googleapis.com/libr...     levantelab   
2355  https://levantelab.storage.googleapis.com/libr...     levantelab   
2356  https://levantelab.storage.googleapis.com/libr...     levantelab   
2385  https://levantelab.storage.googleapis.com/libr...     levantelab   
2722  https://levantelab.storage.googleapis.com/libr...     levantelab   
3053  https://levantelab.storage.googleapis.com/libr...     levantelab   
3054  https://levantelab.storage.googleapis.com/libr...     levantelab   
3055  https://levantelab.storage.googleapis.com/libr...     levantelab   

     video_url_ext                                          fixed_url  
0              mp4  https://levantelab.storage.googleapis.com/libr...  
1              mp4  https://levantelab.storage.googleapis.com/libr...  
2              mp4  https://levantelab.storage.googleapis.com/libr...  
3              mp4  https://levantelab.storage.googleapis.com/libr...  
5              mp4  https://levantelab.storage.googleapis.com/libr...  
6              mp4  https://levantelab.storage.googleapis.com/libr...  
7              mp4  https://levantelab.storage.googleapis.com/libr...  
10   

Sometimes there is multiple videos for one label.

In [ ]:
def number_in_label(label):
    for char in label:
        if char.isdigit():
            return True
    return False


sb_csv["number_in_label"] = sb_csv["label"].apply(number_in_label)
sb_csv["number_in_label"].value_counts()

number_in_label
False    2509
True      575
Name: count, dtype: int64

In the label, these are denoted with:
- {label}-1
- {label}-2

In the URL, these are denoted with:
- /{label}1
- /{label}2

In [ ]:
n = 5
for i, row in sb_csv[sb_csv["number_in_label"] == True].sample(n).iterrows():
    print(row["label"])
    print(row["scraped_url"])
    print()


BRÓCOLIS-2
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/BR%C3%93COLIS2.mp4

YOUTUBE-2
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/YOUTUBE2.mp4

BATISTA-2
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/BATISTA2.mp4

RÁPIDO-2
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/R%C3%81PIDO2.mp4

ALFACE-2
https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/ALFACE2.mp4



Remake levantelab URLs

In [ ]:
def remake_levantelab_url(label):
    if number_in_label(label):
        # ESTADOS-UNIDOS-DA-AMÉRICA-2  -> ESTADOS-UNIDOS-DA-AMÉRICA2
        label = label[:-2] + label[-1]
    url = f"https://videos.nals.cce.ufsc.br/SignBank/Vídeos/{label}.mp4"
    return url


remake_levantelab_url("ESTADOS-UNIDOS-DA-AMÉRICA-2")

'https://videos.nals.cce.ufsc.br/SignBank/Vídeos/ESTADOS-UNIDOS-DA-AMÉRICA2.mp4'

In [ ]:
# apply remake_levantelab_url to the levantelab entries
sb_csv.loc[sb_csv["video_url_root"] == "levantelab", "fixed_url"] = sb_csv.loc[
    sb_csv["video_url_root"] == "levantelab", "label"
].apply(remake_levantelab_url)

In [ ]:
n = 5
for i, row in sb_csv[sb_csv["video_url_root"] == "levantelab"].sample(n).iterrows():
    print(row["label"])
    print(row["scraped_url"])
    print(row["fixed_url"])
    print()

VEGETARIANO
https://levantelab.storage.googleapis.com/libras/gerais/fUObXisMnCnMxh2MHGUHukqDYiz1BhHuVBMCbxOG.mp4
https://videos.nals.cce.ufsc.br/SignBank/Vídeos/VEGETARIANO.mp4

ESTADOS-UNIDOS-DA-AMÉRICA-2
https://levantelab.storage.googleapis.com/libras/gerais/L66RywqgLQWoYLuetwHLLostRXpQCLgfJKRD5T7m.mp4
https://videos.nals.cce.ufsc.br/SignBank/Vídeos/ESTADOS-UNIDOS-DA-AMÉRICA2.mp4

ABACAXI-2
https://levantelab.storage.googleapis.com/libras/gerais/h7ZOyw1GrUWJXxFwU8oJ997sAQE2SY3oyJ9hxI2s.mp4
https://videos.nals.cce.ufsc.br/SignBank/Vídeos/ABACAXI2.mp4

ACUSAR
https://levantelab.storage.googleapis.com/libras/gerais/SSQokxkM9jDJWLIJrRVg1SlnehJD6jq4cFE1c4OJ.mp4
https://videos.nals.cce.ufsc.br/SignBank/Vídeos/ACUSAR.mp4

ABRIR
https://levantelab.storage.googleapis.com/libras/gerais/GIPQ5FwLKp0k9XdAk5ajTtR13qVuxYdqzfsg83xV.mp4
https://videos.nals.cce.ufsc.br/SignBank/Vídeos/ABRIR.mp4



### Make `clean_full_metadata_SignBank.csv`

#### Combine labels with numbers
- usually a dash on labels with numbers, e.g. INCLUSÃO-3
- sometimes no dash, e.g. PARA-FRENTE2

In [ ]:
sb_csv["scraped_label"] = sb_csv["label"]
sb_csv["sign_variant"] = 1

In [ ]:
sb_csv[sb_csv["number_in_label"] == True].sample(5)

letter     label                                        scraped_url  \
1709      M     MÃE-3  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
846       C  CHAMAR-3  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1813      M   MESMO-4  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1985      O    OITO-2  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
2980      V   VERDE-3  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

     video_url_root video_url_ext  \
1709         videos           mp4   
846          videos           mp4   
1813         videos           mp4   
1985         videos           mp4   
2980         videos           mp4   

                                              fixed_url  number_in_label  \
1709  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
846   https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1813  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1985  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
2980  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   

     scraped_label  sign_variant  
1709         MÃE-3             1  
846       CHAMAR-3             1  
1813       MESMO-4             1  
1985        OITO-2             1  
2980       VERDE-3             1

In [ ]:
def remove_number_from_label(label):
    # usually a dash on labels with numbers, e.g. INCLUSÃO-3
    # sometimes no dash, e.g. PARA-FRENTE2

    # go from right to left and remove the number
    if label[-1].isdigit():
        label = label[:-1]
    else:
        print(f"no number at the end of '{label}'")
        return
    if label[-1] == "-":
        label = label[:-1]
    return label


(
    remove_number_from_label("INCLUSÃO-3"),
    remove_number_from_label("PARA-FRENTE2"),
    remove_number_from_label("ESTADOS-UNIDOS-DA-AMÉRICA-2"),
    remove_number_from_label("no-number-test-"),
)

no number at the end of 'no-number-test-'


('INCLUSÃO', 'PARA-FRENTE', 'ESTADOS-UNIDOS-DA-AMÉRICA', None)

In [ ]:
sb_csv.loc[sb_csv["number_in_label"] == True, "sign_variant"] = (
    sb_csv.loc[sb_csv["number_in_label"] == True, "scraped_label"].str[-1].astype(int)
)

In [ ]:
sb_csv.loc[sb_csv["number_in_label"] == True, "label"] = sb_csv.loc[
    sb_csv["number_in_label"] == True, "scraped_label"
].apply(remove_number_from_label)

In [ ]:
sb_csv[sb_csv["number_in_label"] == True].sample(5)

letter     label                                        scraped_url  \
843       C     COMER  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
2138      P     PASTA  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1475      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1572      J    JORNAL  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1033      E  ELEVADOR  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

     video_url_root video_url_ext  \
843          videos           mp4   
2138         videos           mp4   
1475         videos           mp4   
1572         videos           mp4   
1033         videos           mp4   

                                              fixed_url  number_in_label  \
843   https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
2138  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1475  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1572  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1033  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   

     scraped_label  sign_variant  
843        COMER-3             3  
2138       PASTA-2             2  
1475    INCLUSÃO-4             4  
1572      JORNAL-2             2  
1033    ELEVADOR-2             2

Add a `signer_number` col

In [ ]:
sb_csv["signer_number"] = 1
# based on an assumption that the videos at `videos.nals...` have the main signer, and `levantelab` have the other one
sb_csv.loc[sb_csv["video_url_root"] == "levantelab", "signer_number"] = 2

#### Checking any exceptions in the processing

In [ ]:
sb_csv[
    sb_csv.duplicated(subset=["label", "sign_variant", "signer_number"], keep=False)
][["label", "scraped_label", "sign_variant", "number_in_label"]].sort_values(by="label")

label scraped_label  sign_variant  number_in_label
1199     FERRO         FERRO             1            False
1298     FERRO         FERRO             1            False
1472  INCLUSÃO      INCLUSÃO             1            False
1477  INCLUSÃO    INCLUSÃO-1             1             True
2957     VAZIO       VAZIO-2             2             True
3056     VAZIO       VAZIO-2             2             True
3028   VITÓRIA       VITÓRIA             1            False
3029   VITÓRIA       VITÓRIA             1            False

- FERRO
    - Two videos of the same sign for FERRO
    - Not a sign_variant
    - -> Just label different signers
- INCLUSÃO
    - has 6 different sign_variant
    - but labelled differently by SignBank (first variant labelled 1 instead of 2)
    - -> change INCLUSÃO sign_variant to 6
- VAZIO
    - Two videos of the same sign for VAZIO-2
    - Not a sign_variant
    - -> Just label different signers
- VITÓRIA
    - Two videos of different signs for VITÓRIA
    - -> Label one as a sign_variant


Unchecked assumption is that the videos at `videos.nals...` have the main signer, and `levantelab` have the other one`

Ferro

In [ ]:
sb_csv.loc[1298, "signer_number"] = 2

Inclusao

In [ ]:
sb_csv.loc[sb_csv["label"] == "INCLUSÃO"]

letter     label                                        scraped_url  \
1472      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1473      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1474      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1475      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1476      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
1477      I  INCLUSÃO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

     video_url_root video_url_ext  \
1472         videos           mp4   
1473         videos           mp4   
1474         videos           mp4   
1475         videos           mp4   
1476         videos           mp4   
1477         videos           mp4   

                                              fixed_url  number_in_label  \
1472  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   
1473  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1474  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1475  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1476  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
1477  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   

     scraped_label  sign_variant  signer_number  
1472      INCLUSÃO             1              1  
1473    INCLUSÃO-5             5              1  
1474    INCLUSÃO-2             2              1  
1475    INCLUSÃO-4             4              1  
1476    INCLUSÃO-3             3              1  
1477    INCLUSÃO-1             1              1

In [ ]:
sb_csv.loc[1472, "sign_variant"] = 6

Vazio

In [ ]:
sb_csv.loc[sb_csv["scraped_label"] == "VAZIO-2"]

letter  label                                        scraped_url  \
2957      V  VAZIO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
3056      V  VAZIO  https://objetos.sites.ufsc.br/libras/gerais/lI...   

     video_url_root video_url_ext  \
2957         videos           mp4   
3056        objetos           mov   

                                              fixed_url  number_in_label  \
2957  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
3056  https://objetos.sites.ufsc.br/libras/gerais/lI...             True   

     scraped_label  sign_variant  signer_number  
2957       VAZIO-2             2              1  
3056       VAZIO-2             2              1

In [ ]:
sb_csv.loc[sb_csv["scraped_label"] == "VAZIO-2"].scraped_url.values

array(['https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/VAZIO2.mp4',
       'https://objetos.sites.ufsc.br/libras/gerais/lImvYVGMUUeodhH3TuKWk0s2SBGe414Buvi18NsU.mov'],
      dtype=object)

In [ ]:
sb_csv.loc[3056, "signer_number"] = 2

Vitoria

In [ ]:
sb_csv.loc[sb_csv["scraped_label"] == "VITÓRIA"]

letter    label                                        scraped_url  \
3028      V  VITÓRIA  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
3029      V  VITÓRIA  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

     video_url_root video_url_ext  \
3028         videos           mp4   
3029         videos           mp4   

                                              fixed_url  number_in_label  \
3028  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   
3029  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   

     scraped_label  sign_variant  signer_number  
3028       VITÓRIA             1              1  
3029       VITÓRIA             1              1

In [ ]:
sb_csv.loc[sb_csv["scraped_label"] == "VITÓRIA"].scraped_url.values

array(['https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/VIT%C3%93RIA.mp4',
       'https://videos.nals.cce.ufsc.br/SignBank/V%C3%ADdeos/VIT%C3%93RIA-ES.mp4'],
      dtype=object)

In [ ]:
sb_csv.loc[3029, "sign_variant"] = 2

Check the duplicates are fixed

In [ ]:
sb_csv[
    sb_csv.duplicated(subset=["label", "sign_variant", "signer_number"], keep=False)
][
    ["label", "scraped_label", "sign_variant", "number_in_label", "signer_number"]
].sort_values(by="label")

Empty DataFrame
Columns: [label, scraped_label, sign_variant, number_in_label, signer_number]
Index: []

#### Finalise `clean_full_metadata_SignBank.csv` and save

In [199]:
sb_csv.sample(5)

letter        label                                        scraped_url  \
1875      M        MOSCA  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
775       C  COORDENADOR  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
2042      O        OUVIR  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
573       C     CATAPORA  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   
105       A         ALHO  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

     video_url_root video_url_ext  \
1875         videos           mp4   
775          videos           mp4   
2042         videos           mp4   
573          videos           mp4   
105          videos           mp4   

                                              fixed_url  number_in_label  \
1875  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   
775   https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   
2042  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   
573   https://videos.nals.cce.ufsc.br/SignBank/V%C3%...            False   
105   https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             True   

      scraped_label  sign_variant  signer_number  
1875          MOSCA             1              1  
775   COORDENADOR-2             2              1  
2042          OUVIR             1              1  
573        CATAPORA             1              1  
105          ALHO-2             2              1

In [212]:
sb_csv.columns

Index(['letter', 'label', 'scraped_url', 'video_url_root', 'video_url_ext',
       'fixed_url', 'number_in_label', 'scraped_label', 'sign_variant',
       'signer_number'],
      dtype='object')

In [ ]:
sb_csv[
    [
        "label",
        "scraped_label",
        "fixed_url",
        "scraped_url",
        "sign_variant",
        "signer_number",
        "video_url_ext",
        "video_url_root",
        "number_in_label",
        "letter",
    ]
].to_csv("data/raw/SignBank/clean_full_metadata_SignBank.csv", index=False)

### make `metadata.csv`

In [ ]:
clean_signbank = pd.read_csv("data/raw/SignBank/clean_full_metadata_SignBank.csv")

In [17]:
clean_signbank.head()

label scraped_label                                          fixed_url  \
0      À-TOA         À-TOA  https://videos.nals.cce.ufsc.br/SignBank/Vídeo...   
1    À-VISTA       À-VISTA  https://videos.nals.cce.ufsc.br/SignBank/Vídeo...   
2    ABACAXI       ABACAXI  https://videos.nals.cce.ufsc.br/SignBank/Vídeo...   
3     ABANAR        ABANAR  https://videos.nals.cce.ufsc.br/SignBank/Vídeo...   
4  ABANDONAR     ABANDONAR  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...   

                                         scraped_url  sign_variant  \
0  https://levantelab.storage.googleapis.com/libr...             1   
1  https://levantelab.storage.googleapis.com/libr...             1   
2  https://levantelab.storage.googleapis.com/libr...             1   
3  https://levantelab.storage.googleapis.com/libr...             1   
4  https://videos.nals.cce.ufsc.br/SignBank/V%C3%...             1   

   signer_number video_url_ext video_url_root  number_in_label letter  
0              2           mp4     levantelab            False      A  
1              2           mp4     levantelab            False      A  
2              2           mp4     levantelab            False      A  
3              2           mp4     levantelab            False      A  
4              1           mp4         videos            False      A

In [ ]:
clean_signbank.rename(
    columns={
        "fixed_url": "video_url",
        "scraped_url": "scraped_video_url",
    },
    inplace=True,
)

In [ ]:
clean_signbank["data_source"] = "sb"

In [ ]:
clean_signbank[
    [
        "label",
        "video_url",
        "signer_number",
        "data_source",
        "scraped_label",
        "scraped_video_url",
        "sign_variant",
        "signer_number",
        "video_url_root",
        "video_url_ext",
        "number_in_label",
    ]
].to_csv("data/raw/SignBank/metadata.csv", index=False)

## UFV

In [3]:
ufv_path = "data/raw/UFV"
files = os.listdir(ufv_path)
uf_raw_metadata = pd.DataFrame()

# Import and concat all metadata files
for file in (pbar := tqdm(files)):
    pbar.set_description(f"Processing {file}")
    if (file.endswith(".csv")) and ("Metadata" in file):
        temp_df = pd.read_csv(f"{ufv_path}/{file}")
        temp_df = temp_df.drop_duplicates("video_url", keep="last").reset_index(
            drop=True
        )

        uf_raw_metadata = pd.concat([uf_raw_metadata, temp_df], ignore_index=True)

# Drop duplicates (rows with all the same values)
uf_raw_metadata = uf_raw_metadata.drop_duplicates().reset_index(drop=True)

print("Number of rows in uf_raw_metadata:", uf_raw_metadata.shape[0])
print(
    "Number of unique video URLs in uf_raw_metadata:",
    uf_raw_metadata["video_url"].nunique(),
)

Processing UVF_Metadata_8.csv: 100%|██████████| 10/10 [00:00<00:00, 122.72it/s]

Number of rows in uf_raw_metadata: 945
Number of unique video URLs in uf_raw_metadata: 935


In [4]:
uf_raw_metadata.columns

Index(['Word', 'category', 'example_pt', 'example_libras', 'video_url',
       'hand_gesture'],
      dtype='object')

It says here that there are 945 entries for 935 unique URLs. This means that two different rows have the same URL. Let's check some examples.

In [5]:
print(
    "Word",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[-1]["Word"],
)
print(
    "URL:",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[-1]["video_url"],
)
print(
    "Word",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[-2]["Word"],
)
print(
    "URL:",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[-2]["video_url"],
)

Word Errado
URL: https://sistemas.cead.ufv.br/capes/dicionario/wp-content/uploads/2018/09/enxada.mp4
Word Enxada
URL: https://sistemas.cead.ufv.br/capes/dicionario/wp-content/uploads/2018/09/enxada.mp4


It seems like there are some words that had their URL either extracted or registered wrongly in the dataset.  The strategy used to remove these cases is to check if the word is present anywhere in the URL.

In [6]:
print(
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[0]["Word"]
)
print(
    "URL:",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[0]["video_url"],
)
print(
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[1]["Word"]
)
print(
    "URL:",
    uf_raw_metadata[uf_raw_metadata.duplicated(subset=["video_url"], keep=False)]
    .sort_values("video_url")
    .iloc[1]["video_url"],
)

Veneno
URL: Not Available
Maconha
URL: Not Available


Also, there are some words that returned "Not Available" as URLs. These cases are going to be removed.

In [7]:
uf_raw_metadata["url_available"] = uf_raw_metadata["video_url"].apply(
    lambda x: True if x != "Not Available" else False
)

In [8]:
display(uf_raw_metadata["url_available"].value_counts())
uf_raw_metadata = uf_raw_metadata[uf_raw_metadata["url_available"]]

url_available
True     936
False      9
Name: count, dtype: int64

In [9]:
print("Total number of rows:", uf_raw_metadata.shape[0])
print(
    "Number of duplicated URLs:",
    uf_raw_metadata.duplicated("video_url").sum(),
)
print("Number of unique video URLs:", uf_raw_metadata["video_url"].nunique())

Total number of rows: 936
Number of duplicated URLs: 2
Number of unique video URLs: 934


In [10]:
from unidecode import unidecode


def contains_substring(url, word, min_length=3):
    """Check if the word or parts of it are in the URL"""
    if pd.isna(url) or pd.isna(word):
        return False

    url = url.lower()

    substrings = {
        word[i:j]
        for i in range(len(word))
        for j in range(i + min_length, len(word) + 1)
    }

    return any(substring in url for substring in substrings)


# Normalize the Word column
uf_raw_metadata.loc[:, "decoded_word"] = uf_raw_metadata["Word"].str.lower()
uf_raw_metadata.loc[:, "decoded_word"] = uf_raw_metadata["decoded_word"].apply(
    unidecode  # Remove graphic signals
)
uf_raw_metadata.loc[:, "decoded_word"] = uf_raw_metadata["decoded_word"].str.replace(
    " ", "-"
)

# Use contains_substrings to check if the URLs are matching the words
uf_raw_metadata.loc[:, "word_in_url"] = uf_raw_metadata.apply(
    lambda row: contains_substring(row["video_url"], row["decoded_word"], min_length=3),
    axis=1,
)

uf_raw_metadata.head()

Word category                                        example_pt  \
0      Algoritmo  Números                                               NaN   
1        Binário  Números                                               NaN   
2          Cinco  Números                      Minha casa é a número cinco.   
3  Contabilidade  Números  Tenho vontade de fazer o curso de contabilidade.   
4            Dez  Números                           Tenho mais de dez tios.   

                                  example_libras  \
0                                            NaN   
1                                            NaN   
2                           MINHA CASA NUMERO 5.   
3  EU QUERER FUTURO ESTUDAR CURSO CONTABILIDADE.   
4                                EU TER DEZ TIO.   

                                           video_url  \
0  https://sistemas.cead.ufv.br/capes/dicionario/...   
1  https://sistemas.cead.ufv.br/capes/dicionario/...   
2  https://sistemas.cead.ufv.br/capes/dicionario/...   
3  https://sistemas.cead.ufv.br/capes/dicionario/...   
4  https://sistemas.cead.ufv.br/capes/dicionario/...   

                                        hand_gesture  url_available  \
0                                                NaN           True   
1                                                NaN           True   
2  https://sistemas.cead.ufv.br/capes/dicionario/...           True   
3  https://sistemas.cead.ufv.br/capes/dicionario/...           True   
4  https://sistemas.cead.ufv.br/capes/dicionario/...           True   

    decoded_word  word_in_url  
0      algoritmo         True  
1        binario         True  
2          cinco         True  
3  contabilidade         True  
4            dez         True

In [11]:
print(uf_raw_metadata["word_in_url"].value_counts())
uf_metadata = uf_raw_metadata[uf_raw_metadata["word_in_url"]]

word_in_url
True     919
False     17
Name: count, dtype: int64


In [12]:
uf_metadata[uf_metadata.duplicated(subset=["video_url"], keep=False)].sort_values(
    "video_url"
)

Word   category                              example_pt  \
326  Cogumelo   Natureza  Alguns cogumelos servem como alimento.   
524  Cogumelo  Alimentos  Alguns cogumelos servem como alimento.   

    example_libras                                          video_url  \
326            NaN  https://sistemas.cead.ufv.br/capes/dicionario/...   
524            NaN  https://sistemas.cead.ufv.br/capes/dicionario/...   

    hand_gesture  url_available decoded_word  word_in_url  
326          NaN           True     cogumelo         True  
524          NaN           True     cogumelo         True

The table above shows that Cogumelo (mushroom) was set to two different categories (Natureza:Nature and Alimentos:Food), eventhough they have the same sign (URLs send to same video).

In [13]:
uf_metadata[uf_metadata.duplicated(subset="Word", keep=False)]

Word   category                              example_pt  \
326  Cogumelo   Natureza  Alguns cogumelos servem como alimento.   
524  Cogumelo  Alimentos  Alguns cogumelos servem como alimento.   

    example_libras                                          video_url  \
326            NaN  https://sistemas.cead.ufv.br/capes/dicionario/...   
524            NaN  https://sistemas.cead.ufv.br/capes/dicionario/...   

    hand_gesture  url_available decoded_word  word_in_url  
326          NaN           True     cogumelo         True  
524          NaN           True     cogumelo         True

While checking for possible homographs, it's shown here that cogumelo is the only word that show twice in the dataset.

In [14]:
uf_metadata = uf_metadata.rename(columns={"Word": "label"})

In [15]:
uf_metadata.head()

label category                                        example_pt  \
0      Algoritmo  Números                                               NaN   
1        Binário  Números                                               NaN   
2          Cinco  Números                      Minha casa é a número cinco.   
3  Contabilidade  Números  Tenho vontade de fazer o curso de contabilidade.   
4            Dez  Números                           Tenho mais de dez tios.   

                                  example_libras  \
0                                            NaN   
1                                            NaN   
2                           MINHA CASA NUMERO 5.   
3  EU QUERER FUTURO ESTUDAR CURSO CONTABILIDADE.   
4                                EU TER DEZ TIO.   

                                           video_url  \
0  https://sistemas.cead.ufv.br/capes/dicionario/...   
1  https://sistemas.cead.ufv.br/capes/dicionario/...   
2  https://sistemas.cead.ufv.br/capes/dicionario/...   
3  https://sistemas.cead.ufv.br/capes/dicionario/...   
4  https://sistemas.cead.ufv.br/capes/dicionario/...   

                                        hand_gesture  url_available  \
0                                                NaN           True   
1                                                NaN           True   
2  https://sistemas.cead.ufv.br/capes/dicionario/...           True   
3  https://sistemas.cead.ufv.br/capes/dicionario/...           True   
4  https://sistemas.cead.ufv.br/capes/dicionario/...           True   

    decoded_word  word_in_url  
0      algoritmo         True  
1        binario         True  
2          cinco         True  
3  contabilidade         True  
4            dez         True

In [16]:
uf_metadata.to_csv("data/raw/UFV/metadata.csv")

# Combine into one `metadata_combined.csv`

In [17]:
ne_metadata = pd.read_csv("data/raw/INES/metadata.csv")
sb_metadata = pd.read_csv("data/raw/SignBank/metadata.csv")
vl_metadata = pd.read_csv("data/raw/V-Librasil/metadata.csv")
uf_metadata = pd.read_csv("data/raw/UFV/metadata.csv")

In [18]:
metadata_combined = pd.concat([ne_metadata, sb_metadata, vl_metadata, uf_metadata])
metadata_combined.to_csv("data/raw/combined/metadata_combined_unclean.csv", index=False)